### ISSUE TITLE
"Temporary" commit breaks ANSIBLE_KEEP_REMOTE_FILES for 'localhost'

### ISSUE THREAD

**Issue by willthames** [ansible/ansible#35724]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
`lib/ansible/plugins/action/__init__.py`

##### ANSIBLE VERSION

```
ansible 2.5.0 (devel 40f7f4babf) last updated 2018/02/05 21:35:15 (GMT +1000)
  config file = None
  configured module search path = [u'/Users/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/will/src/ansible/lib/ansible
  executable location = /Users/will/src/ansible/bin/ansible
  python version = 2.7.14 (default, Jan  6 2018, 12:16:16) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```

##### CONFIGURATION
ANSIBLE_KEEP_REMOTE_FILES=1

##### OS / ENVIRONMENT
I assume this isn't OS/python related, but in case it is, I spotted this using brew python on OS X

##### SUMMARY
Since bbd6b8bb42cc8fa48, ANSIBLE_KEEP_REMOTE_FILES no longer works.

##### STEPS TO REPRODUCE
Run any playbook with `ANSIBLE_KEEP_REMOTE_FILES=1` set, along with `-vvv`, and then try and explode the resulting module files.

```yaml

```

##### EXPECTED RESULTS
```
$ /usr/local/opt/python/bin/python2.7 /Users/will/.ansible/tmp/ansible-tmp-1517830651.23-79915697239471/stat.py explode
Module expanded into:
/Users/will/.ansible/tmp/ansible-tmp-1517830651.23-79915697239471/debug_dir
```

##### ACTUAL RESULTS

```
$ /usr/local/opt/python/bin/python2.7 /Users/will/.ansible/tmp/ansible-local-15212xlCfsb/ansible-tmp-1517830723.82-13636234799604/stat.py explode
/usr/local/Cellar/python/2.7.14_2/Frameworks/Python.framework/Versions/2.7/Resources/Python.app/Contents/MacOS/Python: can't open file '/Users/will/.ansible/tmp/ansible-local-15212xlCfsb/ansible-tmp-1517830723.82-13636234799604/stat.py': [Errno 2] No such file or directory
```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/plugins/action/__init__.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/__init__.py)
* [lib/ansible/plugins/action/add_host.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/add_host.py)
* [lib/ansible/plugins/action/aireos.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/aireos.py)
* [lib/ansible/plugins/action/aireos_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/aireos_config.py)
* [lib/ansible/plugins/action/aruba.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/aruba.py)
* [lib/ansible/plugins/action/aruba_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/aruba_config.py)
* [lib/ansible/plugins/action/asa.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/asa.py)
* [lib/ansible/plugins/action/asa_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/asa_config.py)
* [lib/ansible/plugins/action/assemble.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/assemble.py)
* [lib/ansible/plugins/action/assert.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/assert.py)
* [lib/ansible/plugins/action/aws_s3.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/aws_s3.py)
* [lib/ansible/plugins/action/bigip.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/bigip.py)
* [lib/ansible/plugins/action/ce.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/ce.py)
* [lib/ansible/plugins/action/ce_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/ce_config.py)
* [lib/ansible/plugins/action/ce_template.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/ce_template.py)
* [lib/ansible/plugins/action/command.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/command.py)
* [lib/ansible/plugins/action/copy.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/copy.py)
* [lib/ansible/plugins/action/debug.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/debug.py)
* [lib/ansible/plugins/action/dellos10.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/dellos10.py)
* [lib/ansible/plugins/action/dellos10_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/dellos10_config.py)
* [lib/ansible/plugins/action/dellos6.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/dellos6.py)
* [lib/ansible/plugins/action/dellos6_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/dellos6_config.py)
* [lib/ansible/plugins/action/dellos9.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/dellos9.py)
* [lib/ansible/plugins/action/dellos9_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/dellos9_config.py)
* [lib/ansible/plugins/action/enos.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/enos.py)
* [lib/ansible/plugins/action/enos_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/enos_config.py)
* [lib/ansible/plugins/action/eos.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/eos.py)
* [lib/ansible/plugins/action/eos_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/eos_config.py)
* [lib/ansible/plugins/action/fail.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/fail.py)
* [lib/ansible/plugins/action/fetch.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/fetch.py)
* [lib/ansible/plugins/action/fortios_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/fortios_config.py)
* [lib/ansible/plugins/action/group_by.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/group_by.py)
* [lib/ansible/plugins/action/include_vars.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/include_vars.py)
* [lib/ansible/plugins/action/ios.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/ios.py)
* [lib/ansible/plugins/action/ios_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/ios_config.py)
* [lib/ansible/plugins/action/iosxr.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/iosxr.py)
* [lib/ansible/plugins/action/iosxr_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/iosxr_config.py)
* [lib/ansible/plugins/action/ironware.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/ironware.py)
* [lib/ansible/plugins/action/ironware_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/ironware_config.py)
* [lib/ansible/plugins/action/junos.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/junos.py)
* [lib/ansible/plugins/action/junos_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/junos_config.py)
* [lib/ansible/plugins/action/net_banner.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_banner.py)
* [lib/ansible/plugins/action/net_base.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_base.py)
* [lib/ansible/plugins/action/net_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_config.py)
* [lib/ansible/plugins/action/net_interface.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_interface.py)
* [lib/ansible/plugins/action/net_l2_interface.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_l2_interface.py)
* [lib/ansible/plugins/action/net_l3_interface.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_l3_interface.py)
* [lib/ansible/plugins/action/net_linkagg.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_linkagg.py)
* [lib/ansible/plugins/action/net_lldp.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_lldp.py)
* [lib/ansible/plugins/action/net_lldp_interface.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_lldp_interface.py)
* [lib/ansible/plugins/action/net_logging.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_logging.py)
* [lib/ansible/plugins/action/net_ping.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_ping.py)
* [lib/ansible/plugins/action/net_static_route.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_static_route.py)
* [lib/ansible/plugins/action/net_system.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_system.py)
* [lib/ansible/plugins/action/net_user.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_user.py)
* [lib/ansible/plugins/action/net_vlan.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_vlan.py)
* [lib/ansible/plugins/action/net_vrf.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/net_vrf.py)
* [lib/ansible/plugins/action/netconf_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/netconf_config.py)
* [lib/ansible/plugins/action/normal.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/normal.py)
* [lib/ansible/plugins/action/nxos.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/nxos.py)
* [lib/ansible/plugins/action/nxos_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/nxos_config.py)
* [lib/ansible/plugins/action/onyx_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/onyx_config.py)
* [lib/ansible/plugins/action/ops_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/ops_config.py)
* [lib/ansible/plugins/action/package.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/package.py)
* [lib/ansible/plugins/action/patch.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/patch.py)
* [lib/ansible/plugins/action/pause.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/pause.py)
* [lib/ansible/plugins/action/raw.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/raw.py)
* [lib/ansible/plugins/action/script.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/script.py)
* [lib/ansible/plugins/action/service.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/service.py)
* [lib/ansible/plugins/action/set_fact.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/set_fact.py)
* [lib/ansible/plugins/action/set_stats.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/set_stats.py)
* [lib/ansible/plugins/action/shell.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/shell.py)
* [lib/ansible/plugins/action/sros.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/sros.py)
* [lib/ansible/plugins/action/sros_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/sros_config.py)
* [lib/ansible/plugins/action/synchronize.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/synchronize.py)
* [lib/ansible/plugins/action/telnet.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/telnet.py)
* [lib/ansible/plugins/action/template.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/template.py)
* [lib/ansible/plugins/action/unarchive.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/unarchive.py)
* [lib/ansible/plugins/action/vyos.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/vyos.py)
* [lib/ansible/plugins/action/vyos_config.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/vyos_config.py)
* [lib/ansible/plugins/action/wait_for_connection.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/wait_for_connection.py)
* [lib/ansible/plugins/action/win_copy.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/win_copy.py)
* [lib/ansible/plugins/action/win_reboot.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/win_reboot.py)
* [lib/ansible/plugins/action/win_template.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/win_template.py)
* [lib/ansible/plugins/action/win_updates.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/win_updates.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @Qalthos @ganeshrn @gundalow @jmighion @kedarX @ogenstad @paulquack @privateip @rcarrillocruz @samerd @trishnaguha
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by bcoca** [comment]
I cannot reproduce:
```
 ANSIBLE_KEEP_REMOTE_FILES=1 ansible -m stat -a 'path=/etc/hosts' -vvvv testhost
```
from output:
```
<127.0.0.3> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/bcoca/.ansible/cp/047663f5ee 127.0.0.3 '/bin/sh -c '"'"'chmod u+x /home/bcoca/.ansible/tmp/ansible-tmp-1517841571.0417922-162025592802770/ /home/bcoca/.ansible/tmp/ansible-tmp-1517841571.0417922-162025592802770/stat.py && sleep 0'"'"''  
```
and then:

```
#>stat /home/bcoca/.ansible/tmp/ansible-tmp-1517841571.0417922-162025592802770/stat.py
  File: /home/bcoca/.ansible/tmp/ansible-tmp-1517841571.0417922-162025592802770/stat.py
  Size: 76498           Blocks: 152        IO Block: 4096   regular file
Device: 803h/2051d      Inode: 9701360     Links: 1
Access: (0700/-rwx------)  Uid: ( 1000/   bcoca)   Gid: ( 1000/   bcoca)
Access: 2018-02-05 09:39:32.190852388 -0500
Modify: 2018-02-05 09:39:32.193852388 -0500
Change: 2018-02-05 09:39:32.201852387 -0500
 Birth: -
```
And when I try w/o the setting, there is no tmp dir.

**Comment by bcoca** [comment]
actually I found the opposite bug, remote temp files never get removed in devel. 389f4ef1fb71629a3389306e6324ea5b118275ff

**Comment by willthames** [comment]
Glad I helped find a (different) bug! I suspect this must be OS X related somehow if you can't reproduce it - I'd have thought I'd have noticed this earlier if it was on all platforms, but OS X is my hobby development environment.

I bisected where it breaks to that very commit, so it's definitely related, but I'll have to investigate this harder later as it's clearly an edge case of some kind.

**Comment by bcoca** [comment]
so reverting this #35733, fixes it for me on other platforms, don't have OS X but it seems really wrong if so.

**Comment by willthames** [comment]
It seems that the temp files are now cleaned up on OS X when ansible or ansible-playbook finishes, and not when the task finishes.

**Comment by willthames** [comment]
bin/ansible runs `shutil.rmtree(C.DEFAULT_LOCAL_TMP)` at the end - it's this that cleans up the remote files.

I think before the change in question, the remote files directory was under something other than `C.DEFAULT_LOCAL_TMP`

Before: ` /Users/will/.ansible/tmp/ansible-tmp-1517917593.4-271581342652258/stat.py`

After: `/Users/will/.ansible/tmp/ansible-local-174741XInKU/ansible-tmp-1517917518.3-107087668300061/stat.py`

where I think `C.DEFAULT_LOCAL_TMP` would be something like `/Users/will/.ansible/tmp/ansible-local-174741XInKU` in both cases, so `shutil.rmtree` would delete the latter but not the former.

The only real question other than how to fix this is why does this *not* happen on Linux?

**Comment by bcoca** [comment]
so those are 'local files' not really remote, which has been removed on controller (or attempted to) for a while now.

My tests were with 'remotes', let my try using local to see if there is some short circuit that is imposing the 'local tmp' setting vs the 'remote tmp' setting.

**Comment by bcoca** [comment]
found the culprit `  new_host.set_variable("ansible_remote_tmp", C.DEFAULT_LOCAL_TMP)` 
this is new var added to the 'implicit localhost' ... sooo .. `ANSIBLE_KEEP_REMOTE_FILES` "still works" but the 'local cleanup' is tacked on top ... i'll remove the setting and we should be back to normal.

**Comment by RickS-C137** [comment]
When will this be fixed? 
I'm running local unit tests for my playbook on my Mac and for troubleshooting need to keep the temporary files created by Ansible. Sadly the environment variable ANSIBLE_KEEP_REMOTE_FILES=1 is ignored and all files are deleted, always. So it would be very helpful to finally fix this bug! It has been reported in another issue almost a year ago!
